### PR TITLE
Add blog post explaining how to manually test

### PR DESCRIPTION
**Pull request by anthonydahanne**
## Checklist

* [X] I have viewed, signed, and submitted the Contributor License Agreement.
* [] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [X] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [X] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+Suppose that you started working on a new feature for the [`spring-boot`](https://github.com/paketo-buildpacks/spring-boot/) buildpack...
+
+## Unit tests first 
+
+Of course, during your development, you will make sure that you did not break any tests, regularly running `go test -v ./...` but you would also add new tests covering your new code; the classic combination `go test -v ./... -coverprofile=coverage.out` followed with `go tool cover -html=coverage.out` will allow you to verify you did not forget to test some errors or edge cases.
+
+## Build and try the buildpack out before publishing it
+
+If the buildpack you are working on provides some integration tests (or end to end tests), I suggest you try and add some use cases to those tests.
```
Comment (dmikusa) [inline content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:18]: Not sure if you want to explain our general testing philosophy with Java buildpacks, I have had questions about it, so it might be worth covering.

1. Test all of the buildpack logic in unit tests. This usually requires tests for individual functions/structs, tests for specific layer providers, and then tests for build and detect.
2. Component buildpacks don't have integration tests. It's difficult to test them in integration because they are usually designed to work together with other buildpacks.
3. Composite buildpacks should have integration tests. This does not need to test every individual feature, because those are covered by 1.) but it should test cross-buildpack compatibility.
4. The Paketo samples offer an additional backstop with additional smoke testing.

While this is the philosophy, we're not quite there. There are a couple of component buildpacks with integration tests (probably should go away) and we've yet to rollout integration tests in the composites (although we approved an RFC to do so).

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+Just run the script and then check you have your buildpacks `bin` and `detect` executable files generated:
+
+```shell
+> scripts/build.sh
+> ls bin                                                                                                                                 
+build  detect helper main
+```
+
+### Mac arm64 note
+If you're using a Mac arm64 (M1, M2, etc.), you may need to update `scripts/build.sh` to precise you want to build for `amd64` (it will help if you publish your image and then try to run it on a  Linux amd64 host); to do that, just prepend `GOARCH="amd64"` before `go build`; like this:
```
Comment (dmikusa) [inline content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:34]:
```suggestion
If you're using a Mac arm64 (M1, M2, etc.), you may need to update `scripts/build.sh` to specify that you want to build for `amd64` (it will help if you publish your image and then try to run it on a  Linux amd64 host); to do that, just prepend `GOARCH="amd64"` before `go build`; like this:
```

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+If not, then you probably want to try it out.
+
+### Build the Paketo Buildpack
+Most Paketo buildpacks provide a `scripts/build.sh` build script that you can use to build the `detect` and `build` binaries (luckily the required `buildpack.toml` is already at the root of the project).
+
+Just run the script and then check you have your buildpacks `bin` and `detect` executable files generated:
+
+```shell
+> scripts/build.sh
```
Comment (dmikusa) [inline content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:28]: I'm not sure you technically need to, but I would suggest running `create-package` instead of running the script directly. That will take care of setting a version.

There are also some [scripts floating around](https://gist.github.com/dmikusa/8e45bb550f194a0606e516372a2f26b9#file-config-fish-L163) (I think @pivotal-david-osullivan might have a bash version) that wrap all this up and package up a dev buildpack image.

I'm conflicted about recommending one of these scripts because they're not official, but I do feel like they make the process easier. Idk, maybe we could create an official script for this?

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  -b paketo-buildpacks/ca-certificates \
+  -b paketo-buildpacks/bellsoft-liberica \
+  -b paketo-buildpacks/syft \
+  -b paketo-buildpacks/maven \
+  -b paketo-buildpacks/executable-jar \
+  -b paketo-buildpacks/apache-tomcat \
+  -b paketo-buildpacks/apache-tomee \
+  -b paketo-buildpacks/liberty \
+  -b paketo-buildpacks/dist-zip \
+  -b .
```
Comment (dmikusa) [inline content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:112]: TIP: If you prepend `urn:cnb:builder:` to the buildpacks you are not modifying then `pack` will take them from the builder. If you do not, then pack will dynamically add all of those buildpacks to the builder. It is slightly faster to use the ones already in the builder.

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+```shell
+pack build -p ~/workspaces/paketo-buildpacks/samples/java/maven applications/maven-app -b anthonydahanne/java:fix-273
+```
+
+##### Updating the builder itself
+
+Now that we can test that the buildpacks inside the composite work well together, we could go even further creating a Paketo builder that would embed this composite and other buildpacks; in case we want to make sure our buildpack is not breaking non-related buildpacks.
+
+Let's update the [`tiny` builder](https://github.com/paketo-buildpacks/tiny-builder) to include our composite.
```
Comment (dmikusa) [inline content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:214]:
```suggestion
Let's update the [`tiny` builder](https://github.com/paketo-buildpacks/builder-jammy-tiny) to include our composite.
```

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+Tip: Run pack build <image-name> --builder anthonydahanne/tiny-builder:fix-273 to use this builder
+```
+
+And finally use it:
+
+```
+pack build -p ~/workspaces/paketo-buildpacks/samples/java/maven applications/maven-app --builder=anthonydahanne/tiny-builder:fix-273
+```
+
+You'll find in the builders repository ([`tiny`](https://github.com/paketo-buildpacks/tiny-builder), [`base`](https://github.com/paketo-buildpacks/base-builder) and [`full`](https://github.com/paketo-buildpacks/full-builder)) integration test that you could try your builder against; to make sure your changes did not break standard expectations.
```
Comment (dmikusa) [inline content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:243]:
```suggestion
You'll find in the builders repository ([`tiny`](https://github.com/paketo-buildpacks/builder-jammy-tiny), [`base`](https://github.com/paketo-buildpacks/builder-jammy-base) and [`full`](https://github.com/paketo-buildpacks/builder-jammy-full)) integration test that you could try your builder against; to make sure your changes did not break standard expectations.
```

**Review by dmikusa — CHANGES_REQUESTED** [review]
Thanks so much for writing this up. It's long overdue. 🙌

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+---
+
+*While this article is mainly about how to manually test Java family Paketo buildpacks, still, most of the instructions should work fine with other Paketo buildpacks; and even non Paketo buildpacks!*
+
+Suppose that you started working on a new feature for the [`spring-boot`](https://github.com/paketo-buildpacks/spring-boot/) buildpack...
+
+## Unit tests first 
+
+Of course, during your development, you will make sure that you did not break any tests, regularly running `go test -v ./...` but you would also add new tests covering your new code; the classic combination `go test -v ./... -coverprofile=coverage.out` followed with `go tool cover -html=coverage.out` will allow you to verify you did not forget to test some errors or edge cases.
+
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:15]: ah! richgo, at first I wasn't sure why we used that, since to me the output palatability is not really enhanced.

But, ok, why not, maybe it made sense at one point.

And then, I read its [creator notice](https://github.com/kyoh86/richgo#notice-what-i-think-about-richgo) and I started envisioning removing it from everywhere...

All in all, I'd prefer not to mention it - it's 1 more library to know (hurdle for newcomers), not much value (go devs. are used to default tests output I assume), and not future proof at all (the notice)

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md (hunk 2 above)
Comment (anthonydahanne) [reply content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md]: oh nice, I did not think of that, much easier since I would not need to stash that change everywhere!

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md (hunk 1 above)
Comment (anthonydahanne) [reply content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:18]: I'm in favor of 3 & 4: I agree integration test at the component bp level is not really interesting.

Integration test should at least cover high level functionality such as "assume gradle spring boot project produces a runnable native image using the `java-native-image` bp composite"

now thinking about it, the samples smoke test (4) do that already...

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md (hunk 3 above)
Comment (anthonydahanne) [reply content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:28]: you're right about the convenience of using `create-package` , but then again: another tool that needs to be installed when just bash and go tool chain and pack cli are enough and probably installed on any buildpacks users machine.

I don't want to explain in this article the `go install -ldflags="-s -w" github.com/paketo-buildpacks/libpak/cmd/create-package@latest` command which additionally makes those instruction even less suitable for other families buildpacks.

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (dmikusa) [reply content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:15]: I need color to find anything in the `go test` output. That's a pain, that it's not being maintained.

Getting off on a tangent, but I'd support removing that since it's not maintained and we don't really need color in CI.

File: content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md (hunk 3 above)
Comment (dmikusa) [reply content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md:28]: That's fair. It might be less awkward to edit buildpack.toml and put in a version number. Otherwise you get the literal string we use as a placeholder in the output. I don't think that will cause issues, but it looks odd so minimally it would be worth calling out why you see that if you don't suggest editing it.

**Comment by anthonydahanne** [comment]
> @anthonydahanne Could you put a profile photo in `assets/images/authors` using the same name that you used in your author field and add your name with that same author id to the `config.toml`

Done in commit https://github.com/paketo-buildpacks/blog/pull/59/commits/d1e3e2c6ecab5b9b7c52dbde5646fc4f8c3f9582 !

**Review by TisVictress — APPROVED** [review]
